otaclient/app/boot_control/_rpi_boot.py

```diff
@@ -0,0 +1,334 @@
+    slot is the fslabel for each AB rootfs.
+
+    This class provides the following features:
+    1. AB partition detection,
+    2. boot files checking,
+    3. switch boot(tryboot.txt setup and tryboot reboot),
+    4. finalize switching boot.
+
+    Boot files for each slot have the following naming format:
+        <boot_file_fname>_<slot>
+    i.e., config.txt for slot_a will be config.txt_slot_a
+    """
+
+    SLOT_A = cfg.SLOT_A_FSLABEL
+    SLOT_B = cfg.SLOT_B_FSLABEL
+    AB_FLIPS = {
+        SLOT_A: SLOT_B,
+        SLOT_B: SLOT_A,
+    }
+    SEP_CHAR = "_"
+
+    def __init__(self) -> None:
+        self.system_boot_path = Path(cfg.SYSTEM_BOOT_MOUNT_POINT)
+        if not (
+            self.system_boot_path.is_dir()
+            and CMDHelperFuncs.is_target_mounted(self.system_boot_path)
+        ):
+            _err_msg = "system-boot is not presented or not mounted!"
+            logger.error(_err_msg)
+            raise BootControlInitError(_err_msg)
+        self._init_slots_info()
+        self._init_boot_files()
+
+    def _init_slots_info(self):
+        """Get current/standby slots info."""
+        try:
+            self._active_slot_dev = CMDHelperFuncs.get_dev_by_mount_point(
+                cfg.ACTIVE_ROOTFS_PATH
+            )
+            self._active_slot = CMDHelperFuncs.get_fslabel_by_dev(self._active_slot_dev)
+            self._standby_slot = self.AB_FLIPS[self._active_slot]
+            self._standby_slot_dev = CMDHelperFuncs.get_dev_by_fslabel(
+                self._standby_slot
+            )
+        except Exception as e:
+            _err_msg = "failed to detect AB partition"
+            logger.error(_err_msg)
+            raise BootControlInitError(_err_msg) from e
+
+    def _init_boot_files(self):
+        """Check the availability of boot files.
+
+        The following boot files will be checked:
+        1. config.txt_<slot_suffix> (required)
+        2. cmdline.txt_<slot_suffix> (required)
+        3. vmlinuz_<slot_suffix>
+        4. initrd.img_<slot_suffix>
+
+        If any of the required files are missing, BootControlInitError will be raised.
+        In such case, a reinstall/setup of AB partition boot files is required.
+        """
+        # boot file
+        self.config_txt = self.system_boot_path / cfg.CONFIG_TXT
+        self.tryboot_txt = self.system_boot_path / cfg.TRYBOOT_TXT
+
+        # active slot
+        self.config_txt_active_slot = (
+            self.system_boot_path / f"{cfg.CONFIG_TXT}{self.SEP_CHAR}{self.active_slot}"
+        )
+        if not self.config_txt_active_slot.is_file():
+            _err_msg = f"missing {self.config_txt_active_slot=}"
+            logger.error(_err_msg)
+            raise BootControlInitError(_err_msg)
+        self.cmdline_txt_active_slot = (
+            self.system_boot_path
+            / f"{cfg.CMDLINE_TXT}{self.SEP_CHAR}{self.active_slot}"
+        )
+        if not self.cmdline_txt_active_slot.is_file():
```

Comment:
I have a quesiton, when and how is this file is generated?

Reply:
config.txt and cmdline.txt for each slot are expected to be installed by the USB installer,  when we initially install autoware image and create AB partition on raspberry.

Reply:
Also, cmdline.txt and config.txt for each slot should not be changed after they are installed, otaclient will copy and replace config.txt_\<slot\> as new config.txt atomically when switching boot.